BUG: issubdtype is inconsistent on types and dtypes

eric-wieser opened the pull request:
Previously, issubdtype(x, y) and issubdtype(dtype(x), dtype(y)) could give different results

Fixes #9480 and #9506 and #2334 (duped by #5711 and #3218)

This deprecates all of the following strange behaviors, with the intent in the distant future to invert the results:

```python
>>> np.issubdtype(np.float32, 'float64')
True
>>> np.issubdtype(np.float32, 'f8')
True
>>> np.issubdtype(np.int32, basestring)
True
>>> np.issubdtype(np.int32, 'int64')
True
>>> np.issubdtype(np.unicode_, 'void')
True
>>> np.issubdtype(np.int32, int)          # should have been spelt np.integer
True
>>> np.issubdtype(np.float32, float)      # should have been spelt np.floating
True
>>> np.issubdtype(np.complex64, complex)  # should have been spelt np.complexfloating
True
>>> np.issubdtype(np.float32, "float")
True
>>> np.issubdtype(np.float64, "f")
True
```

eric-wieser commented:
`test_numerictypes` seemed like the obvious place to put a test for `numerictypes.py`.

However, if you actually look at [that file](https://github.com/eric-wieser/numpy/blob/b5c73ab33d5c1de16244132b252db01affe8d932/numpy/core/tests/test_numerictypes.py), it seems to not be related to numerictypes.py at all

charris commented:
Yeah, the  numpy type hierarchy is a mess. Not conceptually, but as implemented. There are things like
```
In [3]: dtype(np.integer)
Out[3]: dtype('int64')
```
that, IMHO, should raise errors. Perhaps we should make things like `np.integer` abstract classes now that those are available. The main problem with cleaning things up will be the number of backward compatibility problems that will surely arise. The issue should probably be raised on the mailing list because of its potential impact.

eric-wieser commented:
I've thought about this patch some more, and decided the best approach is just to fix dtype inputs, and `FutureWarning` in all other cases

eric-wieser commented:
>  Perhaps we should make things like np.integer abstract classes 

They already are not creatable - `np.integer(1)` fails. Problem is that dtypes don't care whether their type has a constructor.

eric-wieser committed:
BUG: Be explicit about allowing all integer types in matrix exponentiation

Fixes gh-9506, unsigned exponentiation

eric-wieser committed:
MAINT: Stop using the undocumented coercion-then-downcast feature of subdtype

eric-wieser committed:
MAINT: Simplify test for string-likes

eric-wieser commented:
@charris: See updated PR. Apart from the DEP commit, which I can split off if needed, I don't think this needs a mailing list consultation

eric-wieser commented on numpy/core/numerictypes.py:
```diff
@@ -751,14 +752,20 @@ def issubdtype(arg1, arg2):
+            # float, and similar less obvious things, such as np.generic from
+            # basestring
+            mro = arg2.mro()
+            arg2 = mro[1] if len(mro) > 1 else mro[0]
```
Note the only way that `len(mro) == 1` is if `arg2 is object`, which should never happen - but we had it before too, so...

eric-wieser commented on numpy/core/numerictypes.py:
```diff
@@ -751,14 +752,41 @@ def issubdtype(arg1, arg2):
-    return issubclass(dtype(arg1).type, val)
+
+    if not issubclass_(arg1, generic):
+        arg1 = dtype(arg1).type
```
Needed to fix `np.issubdtype(np.number, np.inexact)`

eric-wieser committed:
BUG: abstract types did not compare correctly with issubdtype

eric-wieser committed:
BUG: dtypes are ignored and downcast in issubdtype

Fixes #9480

eric-wieser committed:
DOC: Make example work in python 3

eric-wieser committed:
DEP: Deprecate the coercion of dtype-likes to their superclass

In many cases, this coercion is surprising, or would be if the user knew about it:

* [('a', int)] -> np.flexible
* str - > str (!) - not even a numpy type
* 'float32' -> np.floating (discards size)
* int -> np.signed_integer (not np.integer, as is usually meant)

eric-wieser commented on numpy/core/numerictypes.py:
```diff
@@ -745,20 +746,46 @@ def issubdtype(arg1, arg2):
+            warnings.warn(
+                "Conversion of the second argument of issubdtype from `{raw}` "
+                "to `{abstract} is deprecated. In future, it will be treated as "
+                "`{concrete} == np.dtype({raw}).type`.".format(
```
Example message:
```
In [1]: np.issubdtype(complex, bytearray)
FutureWarning: Conversion of the second argument of issubdtype from `bytearray` to `np.generic is deprecated. In future, it will be treated as `np.object_ == np.dtype(bytearray).type`.
  FutureWarning
Out[1]: True
```

eric-wieser commented:
A quick search reveals that matplotlib makes the `issubdtype(..., int)` mistake in some rarely-used code (combined with the false assumption that `np.int != int`), and scipy makes a few of these too.

This will also catch out people who wrote `np.float` but meant `np.floating`, which scipy is guilty of

charris commented:
LGTM. Could you add an entry in the release notes under `Future Changes`.

eric-wieser commented:
Happy to rebase this myself later.

charris commented:
The test queue looks stuffed. Yeah, go ahead and fix things up and add the release note.

charris commented:
Close and reopen to see if the tests restart.

eric-wieser commented:
Ok, all set, I think.

charris commented:
Thanks Eric.
